DESIGN.md

```diff
@@ -1,31 +1,379 @@
+
+Boolean options without a value are parsed as `true`
+
+For example, parsing command
+
+`rabbitmqctl list_queues --vhost my_vhost -t 10 --formatter=json name pid --quiet`
+
+Will result with unnamed arguments list `["list_queues", "name", "pid"]`
+and named options map `%{vhost: "my_vhost", timeout: 10, quiet: true}`
+
+
+### Usage help
+
+Usage is shown when the CLI is called without any arguments, or if there are some
+problems parsing arguments.
+
+In that cases exit code is `64`
+
+If you want to show usage and return `0` exit code run `help` command
+
+`rabbitmqctl help`
+
+Each tool (`rabbitmqctl`, `rabbitmq-plugins`) shows it's scope of commands (see [Command scopes](#command-scopes))
+
+## Command behaviour
+
+Each command is implemented as an Elixir (or Erlang) module. Command module should
+implement `RabbitMQ.CLI.CommandBehaviour` behaviour.
+
+Behaviour summary:
+
+Following functions MUST be implemented in a command module:
+
+`usage() :: String.t | [String.t]`
+
+Command usage string, or several strings (one per line) to print in command listing in usage.
+Typically looks like `command_name [arg] --option=opt_value`
+
+`flags() :: [Atom.t]`
+
+List of supported option names (e.g. `[:vhost, :timeout]`)
+Error is printed if command doesn't support specified options.
+Global options (except `:vhost`) is ignored.
+Most of the time the keys from the `switches()` callback can be used.
+
+`switches() :: Keyword.t`
+
+Keyword list of switches (argument names and types) for the command.
+
+For example: `def switches(), do: [offline: :boolean, time: :integer]` will
+parse `--offline --time=100` arguments to `%{offline: true, time: 100}` options.
+
+This switches are added to global switches (see [Arguments parsing](#arguments-parsing))
+
+`aliases() :: Keyword.t`
+
+Keyword list of argument names one-letter aliases.
+For example: `[o: :offline, t: :timeout]`
+(see [Arguments parsing](#arguments-parsing))
+
+`banner(arguments :: List.t, options :: Map.t) :: String.t`
+
+Banner to print before the command execution.
+Ignored if argument `--quiet` is specified.
+
+`merge_defaults(arguments :: List.t, options :: Map.t) :: {List.t, Map.t}`
+
+Merge default values for arguments and options (named arguments).
+Returns a tuple with effective arguments and options, that will be passed to `validate/2` and `run/2`
+
+`validate(arguments :: List.t, options :: Map.t) :: :ok | {:validation_failure, Atom.t | {Atom.t, String.t}}`
+
+Validate effective arguments and options.
+
+If function returns `{:validation_failure, err}`
+CLI will print usage to `stderr` and exit with non-zero exit code (typically 64).
+
+`run(arguments :: List.t, options :: Map.t) :: run_result :: any`
+
+Run command. This function usually calls RPC on broker.
+
+`output(run_result :: any, options :: Map.t) :: :ok | {:ok, output :: any} | {:stream, Enum.t} | {:error, ExitCodes.exit_code, [String.t]}`
+
+Cast the return value of `run/2` command to formattable value and exit code.
```

Comment:
Cast the return value of `run/2` command to formattable value and exit code => Cast the return value of the `run/2` callback to a formattable value and exit code